### PR TITLE
MULE-11191: Fix: FTP reconnect Notifier is not working

### PR DESCRIPTION
**Pull request by facundovs**
SimpleRetryPolicy and FtpConnector and FtpMessageReceiver was modified.
FtpReconnectionNotifierTestCase and ftp-reconnection-notifier-fails-config.xml were added.

### CODE REVIEW

File: transports/ftp/src/test/java/org/mule/transport/ftp/FtpReconnectionNotifierTestCase.java [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+        });
+    }
+
+    private void testServerConnected() throws Exception
+    {
+        runFlow("main-test");
+        verifyOnSuccessNotification();
+    }
+
+    private void testServerDesconnected()
```
Comment (alepulver) [inline transports/ftp/src/test/java/org/mule/transport/ftp/FtpReconnectionNotifierTestCase.java]: Desconnected -> Disconnected

File: transports/ftp/src/main/java/org/mule/transport/ftp/FtpConnector.java [hunk 2]
```diff
@@ -543,29 +548,50 @@ public void setupFileType(FTPClient client, ImmutableEndpoint endpoint) throws E
...
+            catch (RetryPolicyExhaustedException retryPolicyExhaustedException)
             {
-                throw new ConnectException(e, this);
+                throw retryPolicyExhaustedException;
             }
 
             try
             {
-                OutputStream out = client.storeFileStream(filename);
+                OutputStream out = client[0].storeFileStream(filename);
```
Comment (pablokraan) [inline transports/ftp/src/main/java/org/mule/transport/ftp/FtpConnector.java]: Is ugly to have this code having to access the array. Looks like this try block can be extracted into an utility method receiving the client as a parameter

File: transports/ftp/src/main/java/org/mule/transport/ftp/FtpMessageReceiver.java [hunk 3]
```diff
@@ -124,11 +126,40 @@ protected boolean pollOnPrimaryInstanceOnly()
...
-            client = connector.createFtpClient(endpoint);
-            FTPListParseEngine engine = client.initiateListParsing();
+            retryTemplate.execute(callbackReconnection, this.connector.getMuleContext().getWorkManager());
+        }
+        catch (RetryPolicyExhaustedException retryPolicyExhaustedException)
+        {
+            throw new ConnectException(retryPolicyExhaustedException,this.connector);
+        }
+
+        try
```
Comment (pablokraan) [inline transports/ftp/src/main/java/org/mule/transport/ftp/FtpMessageReceiver.java]: Same here

File: transports/ftp/src/test/java/org/mule/transport/ftp/FtpReconnectionNotifierTestCase.java [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+import org.apache.ftpserver.usermanager.impl.BaseUser;
+import org.apache.ftpserver.usermanager.impl.WritePermission;
+import org.junit.Rule;
+import org.junit.Test;
+
+
+public class FtpReconnectionNotifierTestCase extends FunctionalTestCase
+{
+
+    private final Integer RECONNECTION_ATTEMPTS = 2;
```
Comment (pablokraan) [inline transports/ftp/src/test/java/org/mule/transport/ftp/FtpReconnectionNotifierTestCase.java]: static

File: transports/ftp/src/test/java/org/mule/transport/ftp/FtpReconnectionNotifierTestCase.java [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+    @Rule
+    public SystemProperty countSystemProperty = new SystemProperty("count", RECONNECTION_ATTEMPTS.toString());
+
+    @Rule
+    public SystemProperty pollingFrequencySystemProperty = new SystemProperty("pollingFrequency", POLLING_FREQUENCY.toString());
+
+    private TestReconnectNotifier mockNotifier = null;
+
+
+    protected String getConfigFile()
```
Comment (pablokraan) [inline transports/ftp/src/test/java/org/mule/transport/ftp/FtpReconnectionNotifierTestCase.java]: Override

File: transports/ftp/src/test/java/org/mule/transport/ftp/FtpReconnectionNotifierTestCase.java [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+    {
+        return "ftp-reconnection-notifier-fails-config.xml";
+    }
+
+    @Test
+    public void testNotificationOnReconnectionAttemptFailure() throws Exception
+    {
+        mockNotifier = mock(TestReconnectNotifier.class);
+        Connector c = muleContext.getRegistry().lookupConnector("FTP");
+        assertNotNull(c);
```
Comment (pablokraan) [inline transports/ftp/src/test/java/org/mule/transport/ftp/FtpReconnectionNotifierTestCase.java]: An assertion like this one is not more helpful than getting a NPE during test execution. Remove it

File: transports/ftp/src/test/java/org/mule/transport/ftp/FtpReconnectionNotifierTestCase.java [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+        RetryPolicyTemplate rpf = c.getRetryPolicyTemplate();
+        assertNotNull(rpf);
+        rpf.setNotifier(mockNotifier);
```
Comment (pablokraan) [inline transports/ftp/src/test/java/org/mule/transport/ftp/FtpReconnectionNotifierTestCase.java]: Not useful to add the mock prefix to variable names, in the same way that we do not name variables like "realNotifier".

File: transports/ftp/src/test/java/org/mule/transport/ftp/FtpReconnectionNotifierTestCase.java [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+        MuleFtpServer server = new MuleFtpServer(2221, 1);
```
Comment (pablokraan) [inline transports/ftp/src/test/java/org/mule/transport/ftp/FtpReconnectionNotifierTestCase.java]: Ports should be dynamic not static, as you can't ensure that the port will be available everytime you run the test (use DynamicPort rule)

File: transports/ftp/src/test/java/org/mule/transport/ftp/FtpReconnectionNotifierTestCase.java [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+        server.start();
+        testServerConnected();
+        server.stop();
+        testServerDisconnected();
+    }
+
+    private void verifyOnSuccessNotification()
+    {
+        PollingProber pollingProber = new PollingProber(POLLING_FREQUENCY * 5, POLLING_FREQUENCY);
```
Comment (pablokraan) [inline transports/ftp/src/test/java/org/mule/transport/ftp/FtpReconnectionNotifierTestCase.java]: POLLING_FREQUENCY * 5 -> use RECEIVE_TIMEOUT (is not a problem to use a bigger value for the timeout, because the timeout is completely consumed only when the test fails)

File: transports/ftp/src/test/java/org/mule/transport/ftp/FtpReconnectionNotifierTestCase.java [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+            public String describeFailure()
+            {
+                return "Success notification wasn't triggered";
+            }
+        });
+    }
+
+    private void verifyOnFailureNotification()
+    {
+        PollingProber pollingProber = new PollingProber(POLLING_FREQUENCY * 5, POLLING_FREQUENCY);
```
Comment (pablokraan) [inline transports/ftp/src/test/java/org/mule/transport/ftp/FtpReconnectionNotifierTestCase.java]: Same as before

File: transports/ftp/src/test/java/org/mule/transport/ftp/FtpReconnectionNotifierTestCase.java [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+
+    private void verifyOnSuccessNotification()
+    {
+        PollingProber pollingProber = new PollingProber(POLLING_FREQUENCY * 5, POLLING_FREQUENCY);
+        pollingProber.check(new JUnitProbe()
+        {
+            @Override
+            protected boolean test() throws Exception
+            {
+                verify(mockNotifier, times(1)).onSuccess(any(RetryContext.class));
```
Comment (pablokraan) [inline transports/ftp/src/test/java/org/mule/transport/ftp/FtpReconnectionNotifierTestCase.java]: time(1) is redundant, is the default condition

File: transports/ftp/src/test/java/org/mule/transport/ftp/FtpReconnectionNotifierTestCase.java [hunk 12]
```diff
@@ -0,0 +1,218 @@
...
+
+    private void verifyOnFailureNotification()
+    {
+        PollingProber pollingProber = new PollingProber(POLLING_FREQUENCY * 5, POLLING_FREQUENCY);
+        pollingProber.check(new JUnitProbe()
+        {
+            @Override
+            protected boolean test() throws Exception
+            {
+                verify(mockNotifier, atLeast(1)).onFailure(any(RetryContext.class), any(Throwable.class));
```
Comment (pablokraan) [inline transports/ftp/src/test/java/org/mule/transport/ftp/FtpReconnectionNotifierTestCase.java]: Is not required to have the verify call inside a try/catch?

File: transports/ftp/src/test/java/org/mule/transport/ftp/FtpReconnectionNotifierTestCase.java [hunk 13]
```diff
@@ -0,0 +1,218 @@
...
+    {
+        mockNotifier = mock(TestReconnectNotifier.class);
+        Connector c = muleContext.getRegistry().lookupConnector("FTP");
+        assertNotNull(c);
+        RetryPolicyTemplate rpf = c.getRetryPolicyTemplate();
+        assertNotNull(rpf);
+        rpf.setNotifier(mockNotifier);
+        MuleFtpServer server = new MuleFtpServer(2221, 1);
+        server.start();
+        testServerConnected();
```
Comment (pablokraan) [inline transports/ftp/src/test/java/org/mule/transport/ftp/FtpReconnectionNotifierTestCase.java]: do not use prefix test for utility methods

File: transports/ftp/src/test/java/org/mule/transport/ftp/FtpReconnectionNotifierTestCase.java (hunk 12 above)
Comment (pablokraan) [inline transports/ftp/src/test/java/org/mule/transport/ftp/FtpReconnectionNotifierTestCase.java]: Why at least? Which condition will cause to have more than one?

File: transports/ftp/src/test/java/org/mule/transport/ftp/FtpReconnectionNotifierTestCase.java [hunk 14]
```diff
@@ -0,0 +1,218 @@
...
+    {
+        runFlow("main-test");
+        verifyOnSuccessNotification();
+    }
+
+    private void testServerDisconnected()
+    {
+        try
+        {
+            runFlow("main-test");
```
Comment (pablokraan) [inline transports/ftp/src/test/java/org/mule/transport/ftp/FtpReconnectionNotifierTestCase.java]: What happens if not exception is thrown on runFlow?

File: transports/ftp/src/test/java/org/mule/transport/ftp/FtpReconnectionNotifierTestCase.java [hunk 15]
```diff
@@ -0,0 +1,218 @@
...
+        }
+    }
+
+
+    public static class MuleFtpServer
+    {
+        private FtpServer server;
+        FtpServerFactory serverFactory;
+        ListenerFactory factory;
+        public MuleFtpServer(int port, long delay)
```
Comment (pablokraan) [inline transports/ftp/src/test/java/org/mule/transport/ftp/FtpReconnectionNotifierTestCase.java]: Add extra line between fields and constructor

File: transports/ftp/src/test/java/org/mule/transport/ftp/FtpReconnectionNotifierTestCase.java [hunk 16]
```diff
@@ -0,0 +1,218 @@
...
+        catch (Exception e)
+        {
+            verifyOnFailureNotification();
+        }
+    }
+
+
+    public static class MuleFtpServer
+    {
+        private FtpServer server;
```
Comment (pablokraan) [inline transports/ftp/src/test/java/org/mule/transport/ftp/FtpReconnectionNotifierTestCase.java]: Add fields visibility

File: transports/ftp/src/test/java/org/mule/transport/ftp/FtpReconnectionNotifierTestCase.java [hunk 17]
```diff
@@ -0,0 +1,218 @@
...
+        FtpServerFactory serverFactory;
```
Comment (pablokraan) [inline transports/ftp/src/test/java/org/mule/transport/ftp/FtpReconnectionNotifierTestCase.java]: Looks like fields can be final

File: transports/ftp/src/test/java/org/mule/transport/ftp/FtpReconnectionNotifierTestCase.java [hunk 18]
```diff
@@ -0,0 +1,218 @@
...
+        ListenerFactory factory;
+        public MuleFtpServer(int port, long delay)
+        {
+
+            serverFactory = new FtpServerFactory();
+            factory = new ListenerFactory();
+            factory.setPort(port);
+            serverFactory.addListener("default", factory.createListener());
+            serverFactory.addListener("default", factory.createListener());
+            PropertiesUserManagerFactory userManagerFactory = new PropertiesUserManagerFactory();
```
Comment (pablokraan) [inline transports/ftp/src/test/java/org/mule/transport/ftp/FtpReconnectionNotifierTestCase.java]: Is all this configuration required? I don't see usages of setHomeDirectory, for example, in existing tests. Also existing test do not add a listener, they use a default one

File: transports/ftp/src/test/java/org/mule/transport/ftp/FtpReconnectionNotifierTestCase.java [hunk 19]
```diff
@@ -0,0 +1,218 @@
...
+
+        public void start()
+        {
+            try
+            {
+                server.start();
+            }
+            catch (FtpException e)
+            {
+                e.printStackTrace();
```
Comment (pablokraan) [inline transports/ftp/src/test/java/org/mule/transport/ftp/FtpReconnectionNotifierTestCase.java]: Do not print the exception, wrap the exception on a RuntimeException and throw it

File: transports/ftp/src/test/java/org/mule/transport/ftp/FtpReconnectionNotifierTestCase.java [hunk 20]
```diff
@@ -0,0 +1,218 @@
...
+
+        public void stop()
+        {
+            try
+            {
+                server.stop();
+            }
+            catch (Exception e)
+            {
+                e.printStackTrace();
```
Comment (pablokraan) [inline transports/ftp/src/test/java/org/mule/transport/ftp/FtpReconnectionNotifierTestCase.java]: Same as before

File: transports/ftp/src/test/java/org/mule/transport/ftp/FtpReconnectionNotifierTestCase.java [hunk 21]
```diff
@@ -0,0 +1,218 @@
...
+            }
+        }
+
+
+    }
+
+    public static class TestReconnectNotifier implements RetryNotifier
+    {
+
+        public static int fails = 0;
```
Comment (pablokraan) [inline transports/ftp/src/test/java/org/mule/transport/ftp/FtpReconnectionNotifierTestCase.java]: Make fields volatile

File: transports/ftp/src/test/java/org/mule/transport/ftp/FtpReconnectionNotifierTestCase.java [hunk 22]
```diff
@@ -0,0 +1,218 @@
...
+        }
+
+        @Override
+        public void onSuccess(RetryContext retryContext)
+        {
+            successes++;
+        }
+
+    }
+
```
Comment (pablokraan) [inline transports/ftp/src/test/java/org/mule/transport/ftp/FtpReconnectionNotifierTestCase.java]: Remove extra lines

File: transports/ftp/src/test/java/org/mule/transport/ftp/FtpReconnectionNotifierTestCase.java (hunk 14 above)
Comment (facundovs) [inline transports/ftp/src/test/java/org/mule/transport/ftp/FtpReconnectionNotifierTestCase.java]: Well If reconnection fails an exception is triggered. 
If I don't throw the expcetion, I can't compare the test results.

File: transports/ftp/src/test/java/org/mule/transport/ftp/FtpReconnectionNotifierTestCase.java (hunk 12 above)
Comment (facundovs) [inline transports/ftp/src/test/java/org/mule/transport/ftp/FtpReconnectionNotifierTestCase.java]: Actually, Failures are two because in testServerDisconnected function runFlow is executed again. I'll replace atLeast condition with times(2).

Comment (facundovs) [inline transports/ftp/src/test/java/org/mule/transport/ftp/FtpReconnectionNotifierTestCase.java]: No. It isn't required.

File: transports/ftp/src/test/java/org/mule/transport/ftp/FtpReconnectionNotifierTestCase.java (hunk 14 above)
Comment (pablokraan) [inline transports/ftp/src/test/java/org/mule/transport/ftp/FtpReconnectionNotifierTestCase.java]: I guess you did not understand the previous comment. You are expecting an exception to be thrown, you are checking that on verifyOnFailureNotification. But if no exception is thrown, the test will pass

File: modules/db/src/test/resources/integration/reconnect/derby-db-reconnect-standard-config.xml [hunk 23]
```diff
@@ -7,7 +7,7 @@
 
     <!-- Uses an invalid URL to avoid connecting until the URL is fixed using the reconnect notifier-->
     <db:derby-config name = "badDbConfig" url="jdbc:derby:garbage">
-        <reconnect>
+        <reconnect count="3">
```
Comment (pablokraan) [inline modules/db/src/test/resources/integration/reconnect/derby-db-reconnect-standard-config.xml]: This change does not seem related to the fixed issue

File: core/src/main/java/org/mule/retry/policies/SimpleRetryPolicy.java [hunk 24]
```diff
@@ -84,7 +84,7 @@ protected boolean isApplicableTo(Throwable cause)
     protected boolean isExhausted()
     {
         return count != SimpleRetryPolicyTemplate.RETRY_COUNT_FOREVER
-               && retryCounter.current().get() >= count;
+               && retryCounter.current().get() >= count -1;
```
Comment (pablokraan) [inline core/src/main/java/org/mule/retry/policies/SimpleRetryPolicy.java]: Should not be retryCounter.current().get() > count?

File: transports/ftp/src/main/java/org/mule/transport/ftp/FtpConnector.java [hunk 25]
```diff
@@ -543,58 +548,45 @@ public void setupFileType(FTPClient client, ImmutableEndpoint endpoint) throws E
...
-            catch (Exception e)
+            catch (RetryPolicyExhaustedException retryPolicyExhaustedException)
             {
-                logger.debug("Error getting output stream: ", e);
-                releaseFtp(uri, client);
-                throw e;
+                throw retryPolicyExhaustedException;
             }
+
+            return storeFileStream(client[0],filename,uri);
```
Comment (pablokraan) [inline transports/ftp/src/main/java/org/mule/transport/ftp/FtpConnector.java]: Fix code style

File: transports/ftp/src/test/java/org/mule/transport/ftp/FtpReconnectionNotifierTestCase.java [hunk 26]
```diff
@@ -0,0 +1,212 @@
...
+        return "ftp-reconnection-notifier-fails-config.xml";
+    }
+
+    @Test
+    public void testNotificationOnReconnectionAttemptFailure() throws Exception
+    {
+        notifier = mock(TestReconnectNotifier.class);
+        Connector c = muleContext.getRegistry().lookupConnector("FTP");
+        RetryPolicyTemplate rpf = c.getRetryPolicyTemplate();
+        assertNotNull(rpf);
```
Comment (pablokraan) [inline transports/ftp/src/test/java/org/mule/transport/ftp/FtpReconnectionNotifierTestCase.java]: Remove this assertion
